packages/react/src/xcss-prop/index.ts

```diff
@@ -0,0 +1,143 @@
+import type * as CSS from 'csstype';
+
+import { ac } from '../runtime';
+import type { CSSPseudos } from '../types';
+
+type CSSProperties = Readonly<CSS.Properties<string | number>>;
+
+type XCSSItem<TStyleDecl extends keyof CSSProperties> = {
+  [Q in keyof CSSProperties]: Q extends TStyleDecl
+    ? CompiledPropertyDeclarationReference | string | number
+    : never;
+};
+
+type XCSSPseudos<K extends keyof CSSProperties, TPseudos extends CSSPseudos> = {
+  [Q in CSSPseudos]?: Q extends TPseudos ? XCSSItem<K> : never;
+};
+
+/**
+ * We currently block all at rules from xcss prop.
+ * This needs us to decide on what the final API is across Compiled to be able to set.
+ */
+type XCSSAtRules = {
+  [Q in CSS.AtRules]?: never;
+};
+
+type CompiledPropertyDeclarationReference = {
+  ['__COMPILED_PROPERTY_DECLARATION_REFERENCE_DO_NOT_WRITE_DIRECTLY__']: true;
+};
+
+/**
+ * Used to mark styles that have been flushed through an API as being generated
+ * from Compiled. This is useful when you want other ends of the API to ensure they
+ * take Compiled generated styles and not some arbitrary object.
+ */
+export type CompiledStyles<TObject> = {
+  [Q in keyof TObject]: TObject[Q] extends Record<string, unknown>
+    ? CompiledStyles<TObject[Q]>
+    : CompiledPropertyDeclarationReference;
+};
+
+/**
+ * Please think twice before using this type, you're better off declaring explicitly
+ * what your API should be, for example only defining `color`.
+ *
+ * Use in conjunction with {@link XCSSProp} to allow all properties to be given to
+ * your component.
+ */
+export type XCSSAllProperties = keyof CSSProperties;
```

Comment:
What's the reasoning behind using the word `All` in the type name rather than just `XSSProperties`? Is it to make it clearer you don't want it used?

Reply:
Yeah to make it clear you're opting into everything. It's really only for makers declaring xcss prop, not for internal Compiled usage.